tools/advisor/advisor/db_config_optimizer.py

```diff
@@ -0,0 +1,385 @@
+                        updated_config[sugg.option] = {}
+                    if '.' not in sugg.option:
+                        updated_config[sugg.option][NO_FAM] = new_value
+                    else:
+                        for col_fam in rule.get_trigger_column_families():
+                            updated_config[sugg.option][col_fam] = new_value
+                continue
+            # case: when the option is present in the current configuration
+            if NO_FAM in current_config[sugg.option]:
+                old_value = current_config[sugg.option][NO_FAM]
+                new_value = ConfigOptimizer.apply_action_on_value(
+                    old_value, sugg.action, sugg.suggested_values
+                )
+                if new_value:
+                    if sugg.option not in updated_config:
+                        updated_config[sugg.option] = {}
+                    updated_config[sugg.option][NO_FAM] = new_value
+            else:
+                for col_fam in rule.get_trigger_column_families():
+                    old_value = None
+                    if col_fam in current_config[sugg.option]:
+                        old_value = current_config[sugg.option][col_fam]
+                    new_value = ConfigOptimizer.apply_action_on_value(
+                        old_value, sugg.action, sugg.suggested_values
+                    )
+                    if new_value:
+                        if sugg.option not in updated_config:
+                            updated_config[sugg.option] = {}
+                        updated_config[sugg.option][col_fam] = new_value
+        return current_config, updated_config
+
+    @staticmethod
+    def pick_rule_to_apply(rules, current_rule_name, rules_tried, backtrack):
+        if not rules:
+            print('\nNo more rules triggered!')
+            return None
+        if current_rule_name and not backtrack:
+            for rule in rules:
+                if rule.name == current_rule_name:
+                    return rule
+        for rule in rules:
+            if rule.name not in rules_tried:
+                return rule
+        print('\nAll rules have been exhausted')
+        return None
+
+    @staticmethod
+    def apply_suggestions(
+        triggered_rules,
+        current_rule_name,
+        rules_tried,
+        backtrack,
+        curr_options,
+        suggestions_dict
+    ):
+        curr_rule = ConfigOptimizer.pick_rule_to_apply(
+            triggered_rules, current_rule_name, rules_tried, backtrack
+        )
+        if not curr_rule:
+            return tuple([None]*4)
+        # if a rule has been picked for improving db_config, update rules_tried
+        rules_tried.add(curr_rule.name)
+        # get updated config based on the picked rule
+        curr_conf, updated_conf = ConfigOptimizer.improve_db_config_v2(
+            curr_options, curr_rule, suggestions_dict
+        )
+        conf_diff = DatabaseOptions.get_options_diff(curr_conf, updated_conf)
```

Comment:
What is the purpose of getting the diff? Based on our offline conversion this change can be rolled back, right?

Reply:
Leaving this as it is (as per offline discussion)